R/derive_adlb_params.R

```diff
@@ -0,0 +1,408 @@
+      )
+
+      missing_elements <- setdiff(unit_choices, names(thresholds[[param]]))
+
+      if (length(missing_elements) >= 1L) {
+        cli_abort(
+          message =
+            "The following required elements are missing from {.arg {param}} child
+            of {.arg {arg_name}} argument: {.val {unit_choices}}.",
+          class = class,
+          call = call
+        )
+      }
+    }
+
+    invisible(thresholds)
+  }
+
+  assert_thresholds(prediabetic_thresholds)
+  assert_thresholds(diabetic_thresholds)
+
+  # Initial derivation assigning either NORMOGLYCEMIC or PREDIABETIC status ----
+
+  dataset <- dataset %>%
+    mutate(
+      # Map `prediabetic_thresholds` into a variable
+      PRED_THRESHOLD = purrr::map2_dbl(
+        PARAMCD, !!get_unit_expr,
+        ~ purrr::pluck(prediabetic_thresholds, .x, .y) %||%
+          purrr::pluck(prediabetic_thresholds, .x, .default = NA_real_)
+      )
+    ) %>%
+    derive_param_computed(
+      by_vars = by_vars,
+      parameters = c(hba1c_code, fpg_code),
+      set_values_to = exprs(
+        AVALC = case_when(
+          (
+            !!sym(paste0("AVAL.", hba1c_code)) >= !!sym(paste0("PRED_THRESHOLD.", hba1c_code)) |
+              !!sym(paste0("AVAL.", fpg_code)) >= !!sym(paste0("PRED_THRESHOLD.", fpg_code))
+          )
+          ~ "PREDIABETIC",
+          (
+            !!sym(paste0("AVAL.", hba1c_code)) < !!sym(paste0("PRED_THRESHOLD.", hba1c_code)) |
+              !!sym(paste0("AVAL.", fpg_code)) < !!sym(paste0("PRED_THRESHOLD.", fpg_code))
+          )
+          ~ "NORMOGLYCEMIC",
+          .default = "UNKNOWN"
+        ),
+        !!!set_values_to
+      ),
+      keep_nas = TRUE
+    ) %>%
+    select(-PRED_THRESHOLD)
+
+  # Identify records that meet the criteria for DIABETIC status ----
+
+  # Diabetic status requires a confirmation, so that there should be
+  # two consecutive time points where the results meet the criteria.
+
+  diabetes_confirmed <- dataset %>%
+    filter(
+      PARAMCD %in% c(hba1c_code, fpg_code)
+    ) %>%
+    mutate(
+      # Map `diabetic_thresholds` into a variable
+      D_THRESHOLD = purrr::map2_dbl(
+        PARAMCD, !!get_unit_expr,
+        ~ purrr::pluck(diabetic_thresholds, .x, .y) %||%
+          purrr::pluck(diabetic_thresholds, .x, .default = NA_real_)
+      ),
+      # Change PARAMCD values for convenience
+      PARAMCD = case_match(
+        PARAMCD,
+        {{ hba1c_code }} ~ "HBA1C",
+        {{ fpg_code }} ~ "FPG"
+      )
+    ) %>%
+    # Transpose the results and thresholds to a wide structure
+    tidyr::pivot_wider(
+      id_cols = vars2chr(by_vars),
+      names_from = "PARAMCD",
+      names_glue = "{PARAMCD}.{.value}",
+      values_from = c("AVAL", "D_THRESHOLD")
+    ) %>%
+    # Identify cases where the criteria is met twice consecutively
+    derive_var_joined_exist_flag(
+      dataset = .,
+      dataset_add = .,
+      by_vars = get_admiral_option("subject_keys"),
```

Comment:
Would there be occasions where this would not work? I.e. there is no direct requirements to use admiral options to set the by vars.